sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1201 @@
+        return str(self.serialize())
+
+    def __str__(self):
+        return str(repr(self))
+
+
+class MetadataRequest:
+    """Contains the metadata request info of a consumer.
+
+    Instances of :class:`MethodMetadataRequest` are used in this class for each
+    available method under `metadatarequest.{method}`.
+
+    Consumers-only classes such as simple estimators return a serialized
+    version of this class as the output of `get_metadata_routing()`.
+
+    .. versionadded:: 1.1
+    """
+
+    def __init__(self):
+        for method in METHODS:
+            setattr(self, method, MethodMetadataRequest(name=method))
+
+    def _get_param_names(self, method, original_names, ignore_self=None):
+        """Get the names of all available metadata for a method.
+
+        This method returns the names of all metadata, even the UNREQUESTED
+        ones.
+
+        Parameters
+        ----------
+        method : str
+            The name of the method for which metadata names are requested.
+
+        original_names : bool
+            Controls whether original or aliased names should be returned. If
+            ``True``, aliases are ignored and original names are returned.
+
+        ignore_self : bool
+            Ignored. Present for API compatibility.
+
+        Returns
+        -------
+        names : set of str
+            Returns a set of strings with the names of all parameters.
+        """
+        return getattr(self, method)._get_param_names(original_names=original_names)
+
+    def _get_params(self, *, method, params):
+        """Return the input parameters requested by a method.
+
+        The output of this method can be used directly as the input to the
+        corresponding method as extra props.
+
+        Parameters
+        ----------
+        method : str
+            The name of the method for which the parameters are requested and
+            routed.
+
+        params : dict
+            A dictionary of provided metadata.
+
+        Returns
+        -------
+        params : Bunch
+            A :class:`~utils.Bunch` of {prop: value} which can be given to the
+            corresponding method.
+        """
+        return getattr(self, method)._get_params(params=params)
+
+    def _check_warnings(self, *, method, params):
+        """Check whether metadata is passed which is marked as WARN.
+
+        If any metadata is passed which is marked as WARN, a warning is raised.
+
+        Parameters
+        ----------
+        method : str
+            The name of the method for which the warnings should be checked.
+
+        params : dict
+            The metadata passed to a method.
+        """
+        getattr(self, method)._check_warnings(params=params)
+
+    def serialize(self):
+        """Serialize the object.
+
+        Returns
+        -------
+        obj : dict
+            A serialized version of the instance in the form of a dictionary.
+        """
+        output = {"^type": "request"}
```

Comment:
Why `^`? I might have expected `$`, `@`, `#` or `%`, but I've never seen `^` used... where does this come from?

Reply:
Put that on my lack of experience serializing objects of different types and mixing them. I'm happy to replace `^` with any of the other alternatives (except `@` I guess?). Just need to make it an invalid identifier.